### ISSUE TITLE
Markers scaling behaves differently in different computing environments

### ISSUE THREAD

**Issue by alisterburt** [vispy/vispy#2078]
We noticed some strange behaviour over at napari/napari#2894 - depending on the computing environment our `MarkersVisual` with `scaling=True` does not always scale points as expected. Specific info is in the issue there, but basically I think it behaves as expected on @brisvag's linux box and the scaling is off on my M1 macbook air.

Is this behaviour something people here have seen/have any idea how to debug/work around? For an application I'm working on having markers stay the same size relative to images would be really nice 🙂

**Comment by djhoese** [comment]
Given Apple's deprecation of OpenGL support, could you determine if there are any non Mac systems that this is an issue on?

**Comment by djhoese** [comment]
Oh and can you get it down to a vispy-only example?

**Comment by alisterburt** [comment]
Thanks @djhoese - I'll ask around

Here is a vispy only example 

```Python
import numpy as np
import vispy.app
import vispy.scene
from vispy.scene import Markers, Image

# checkerboard image
image = np.zeros((100, 100))
image[::2, ::2] = 1


# markers
x = np.linspace(4, 94, 10, endpoint=True)
xx, yy = np.meshgrid(x, x)
markers = np.vstack((xx.flatten(), yy.flatten())).transpose()

canvas = vispy.scene.SceneCanvas(keys='interactive', show=True)
view = canvas.central_widget.add_view()

image_visual = Image(cmap='gray')
image_visual.set_data(image)

markers_visual = Markers()
markers_visual.set_data(markers, scaling=True, size=4, face_color='cornflowerblue')

view.add(markers_visual)
view.add(image_visual)

view.camera = vispy.scene.PanZoomCamera(aspect=1)
vispy.app.run()
```

**Comment by brisvag** [comment]
Can confirm that this works fine on my machine (i.e: the marker size to image ration stays constant).

**Comment by asnt** [comment]
I have the same issue with one system.

It seems there is a driver/hardware-dependent limit on the maximum point size.
https://docs.gl/gl2/glPointSize
https://community.khronos.org/t/adjust-gl-points-size/67980/6

```python
>>> GL_ALIASED_POINT_SIZE_RANGE = 33901
>>> from vispy.gloo import gl
>>> size = gl.glGetParameter(GL_ALIASED_POINT_SIZE_RANGE)
>>> print(size)
(1.0, 255.0) # On an intel integrated card.
>>> print(size)
(1.0, 2047.0) # On an nvidia card.
```

**Comment by alisterburt** [comment]
@asnt thanks, this is really useful to know about!

**Comment by maweigert** [comment]
I can confirm this bug (on macOS 11.6 and 12.0.1, both M1) - scaling of points seems to be currently not working at all.

Original example from @alisterburt :


https://user-images.githubusercontent.com/11042162/144241434-92f87654-afe3-4478-ba98-c2fc8765125d.mp4

**Comment by marcel-goldschen-ohm** [comment]
Not sure if this will help as I didn't fully understand it, but here is a link to a potential workaround for the openGL point size limit: https://webglfundamentals.org/webgl/lessons/webgl-qna-working-around-gl_pointsize-limitations-webgl.html

Maybe this will be of help in solving this very annoying issue that points do not scale as expected when zooming.

**Comment by brisvag** [comment]
hey @marcel-goldschen-ohm, thanks for the interest and the link! Yes, I considered doing the same some time ago, but it takes some work and it didn't seem like a super urgent issue. If you or anyone else is interested in picking this up, there is some code that does a similar job (albeit a bit more complex because it was intended for non-square markers) in an [old branch of mine, using geometry shaders](https://github.com/brisvag/vispy/blob/bf93393fed0527ab3ba92567fc5224bf548875fc/vispy/visuals/cylinders.py#L50-L66). I doubt anything works there, unfortunately, but I know it can :P

**Comment by marcel-goldschen-ohm** [comment]
Thanks for the info @brisvag, I'm not sure I have the bandwidth to tackle this but if I do your earlier work should at least provide me with a starting point ;) Cheers!

**Comment by brisvag** [comment]
I'm afraid not :/ see my latest comment above for some more details!

**Comment by LucaMarconato** [comment]
> I have the same issue with one system.
> 
> It seems there is a driver/hardware-dependent limit on the maximum point size.
> [docs.gl/gl2/glPointSize](https://docs.gl/gl2/glPointSize)
> [community.khronos.org/t/adjust-gl-points-size/67980/6](https://community.khronos.org/t/adjust-gl-points-size/67980/6)
> 
> >>> GL_ALIASED_POINT_SIZE_RANGE = 33901
> >>> from vispy.gloo import gl
> >>> size = gl.glGetParameter(GL_ALIASED_POINT_SIZE_RANGE)
> >>> print(size)
> (1.0, 255.0) # On an intel integrated card.
> >>> print(size)
> (1.0, 2047.0) # On an nvidia card.

Also facing the issue on my M1 Mac. From the quoted message it seems like it is a problem also of the older architecture? Is this correct or is only an Apple Silicon problem?

**Comment by brisvag** [comment]
It was a problem on intel macs already. If you run the above code on the M1 do you get similar outputs?

**Comment by LucaMarconato** [comment]
Uh, running this code on I get a segfault
```python
GL_ALIASED_POINT_SIZE_RANGE = 33901
from vispy.gloo import gl
size = gl.glGetParameter(GL_ALIASED_POINT_SIZE_RANGE)
```

@giovp can you please try to run it?

**Comment by giovp** [comment]
same!
```
>>> size = gl.glGetParameter(GL_ALIASED_POINT_SIZE_RANGE)
[1]    94952 segmentation fault  python
```

**Comment by psobolewskiPhD** [comment]
I'm not sure why it segfaults, but to get looking at 
https://github.com/vispy/vispy/blob/c6e334879d5afde7cda29bbf76aee5d797189e6c/vispy/util/config.py#L390
the gl parameters you need a context. Cribbing the code from there, this works for me on my arm64 MBPro.

```
from vispy.app import Canvas, use_app
from vispy.gloo import gl

app = use_app(call_reuse=False)
canvas = Canvas('Test', (10, 10), show=False, app=app)

GL_ALIASED_POINT_SIZE_RANGE = 33901
size = gl.glGetParameter(GL_ALIASED_POINT_SIZE_RANGE)
print(size)
```
It returns (1.0, 64.0)

**Comment by brisvag** [comment]
So, still tiny on latest macs :/

**Comment by psobolewskiPhD** [comment]
Correct, though I thought I saw something that the non-aliased size can be larger. But not sure how to verify.

**Comment by alisterburt** [comment]
@kesterlester I understand your frustration. I'm sure if you wanted to contribute this improvement to vispy to provide the solution you need in napari it would be much appreciated by many.

Please remember that napari is an open source project, historically mostly run on volunteer labour even if there was funding in the napari-sphere :-)

**Comment by kesterlester** [comment]
> [@kesterlester](https://github.com/kesterlester) I understand your frustration. I'm sure if you wanted to contribute this improvement to vispy to provide the solution you need in napari it would be much appreciated by many.
> 
> Please remember that napari is an open source project, historically mostly run on volunteer labour even if there was funding in the napari-sphere :-)

Absolutely. I did not intend my message to mean that others should to this for me, and indeed if I do manage to find a way of doing something like this I would contribute it. 

But rather I just meant to flag up (perhaps not very usefully) that the issue is still a live one. 

Sometimes that helps ? if people are looking at what people care about.

**Comment by kesterlester** [comment]
> [@kesterlester](https://github.com/kesterlester) it seems you actually _want_ the behaviour of the aliasing macOS version. Is that correct?

Indirectly yes.   I will relate why below, since you ask.  But first regarding:

> [@kesterlester](https://github.com/kesterlester)  If that's the case there might be something simpler we can do to (optionally) mimic the aliasing on all platforms. At least that unifies things. This also might be more achievable in napari without a change to vispy.

Here I would say, no: I would recommend a feature is put in that purposely mimics the macos behaviour on all platforms (well, at least not on my account) as (a) it is a bit of a strange behaviour so it's probably only a niche interest, and (b) it looks like is not too tricky to implement it ad-hoc callback that triggers on zoom level changes and resizes sybols dynamically (works in my case since it only has to keep symbols below the size that the OS can manage).  

So rather than having my needs met, I think it would be better encourage focusing any effort [if there was any] on 

(i) removing surprises for new users (e.g. via warnings), and

 (ii) making the already planned/expected behaviour (i.e. symbols are constant data size) consistent across platforms. 

My own needs were probably niche, so don't focus on me! :)

-- 

Back to your first question:

As it happens, in our application it was beneficial for symbols to be a largest (approximately fixed size relative to screen)  when zoomed far out of the image, but then to shrink (relative to the data size) as you zoomed in.

I didn't initially know how symbol sizes were supposed to behave -- but the very first program I wrote happened to set a very large symbol size (~400) much larger than the mac could cope with.  That, given the very large size of our images, and strange way the mac behaves, happened to do more or less exactly what we wanted!   [I.e. when zoomed out 400 data-pixels was within the 60 screen=pixel capability of the mac opengl, so the symbols were large enough and looked good. But then as one zoomed in, the symbols would initially grow a little, before saturating at the max size that mac opengl could handle, at which point the rest of the image would continue to grow while the symbols shrank relative to it. All what we wanted.

I didn't think much about this behaviour when I saw it, as I just assumed that this was "standard" and "intended" rather than unusual.  [ E.g.  if you are in power point or fusion350, say, and you zoom into something you are working on, then the  size of the mouse pointer or the handles on the corners of objects all stay fixed screen size. They don't grow with the zoom level. It looked to me like the points in napari / vispy had just been set to behave in a similar way when zoomed out, but a different way when zoomed in. Perhaps to allow data to be less obscured at high zoom levels. :) ]

So I was just lucky to get something that fitted my needs, and not savvy enough to realise that I was just lucky.

**Comment by TimMonko** [comment]
@kesterlester 
This recent napari PR https://github.com/napari/napari/pull/7732 reminds me of your situation, though perhaps inverse? Either way, it shows how to use the `camera.zoom` to change the size of objects. If I'm reading correctly you want the markers to almost work like unscaled text in napari? (i.e., before the PR)
If this is some use to you it be cool to have this as a contribution to napari itself. I think there is plenty of flexibility for marker customization on the napari side.
To be fair, thinking about this has me feeling a bit lost -- despite making the PR lol
